source_code/main_mcu/src/LOGIC/logic_user.c

```diff
@@ -1703,31 +1748,133 @@ RET_TYPE logic_user_ask_for_credentials_keyb_output(uint16_t parent_address, uin
                     typing_message_to_be_sent->keyboard_type_message.delay_between_types = custom_fs_settings_get_device_setting(SETTINGS_DELAY_BETWEEN_PRESSES);
                     typing_message_to_be_sent->keyboard_type_message.interface_identifier = interface_id;
                     comms_aux_mcu_send_message(typing_message_to_be_sent);
-                    
+
+
+                    /* Wait for typing status */
+                    while(comms_aux_mcu_active_wait(&temp_rx_message, FALSE, AUX_MCU_MSG_TYPE_KEYBOARD_TYPE, FALSE, -1) != RETURN_OK){}
+                    could_type_all_symbols = (BOOL)temp_rx_message->payload_as_uint16[0];
+
+                    /* Rearm DMA RX */
+                    comms_aux_arm_rx_and_clear_no_comms();
+
+                    /* Display warning if some chars were missing */
+                    if ((string_to_key_points_transform_success != RETURN_OK) || (could_type_all_symbols == FALSE))
+                    {
+                        gui_prompts_display_information_on_screen_and_wait(COULDNT_TYPE_CHARS_TEXT_ID, DISP_MSG_WARNING, FALSE);
+                    }
+
+                    /* Set bool */
+                    login_or_password_typed = TRUE;
+
+                }
+                /* Move on */
+                state_machine++;
+            }
+        }
+        else if (state_machine == 3)
+        {
+            /* TOTP provided? */
+            if (temp_cnode.TOTPsecretLen == 0)
+            {
+                /* Message is sent, clear everything */
+                memset(&temp_cnode, 0, sizeof(temp_cnode));
+                if (login_or_password_typed == FALSE)
+                {
+                    return RETURN_NOK;
+                }
+                else
+                {
+                    return RETURN_OK;
+                }
+            }
+
+            /* Ask for TOTP confirmation */
+            cust_char_t totp_prompt[15];
+            memset(totp_prompt, 0, sizeof(totp_prompt));
+            //TODO: Hardcoded text
+            strcpy((char *) totp_prompt, "Type TOTP?");
+            utils_ascii_to_unicode((uint8_t *)totp_prompt, 10);
+            conf_text_2_lines.lines[1] = totp_prompt;
+
+            mini_input_yes_no_ret_te prompt_return = gui_prompts_ask_for_confirmation(2, &conf_text_2_lines, FALSE, TRUE, TRUE);
+
+            /* Approved, back, card removed... */
+            if ((prompt_return == MINI_INPUT_RET_CARD_REMOVED) || (prompt_return == MINI_INPUT_RET_POWER_SWITCH))
+            {
+                /* Message is sent, clear everything */
+                memset(&temp_cnode, 0, sizeof(temp_cnode));
+                return RETURN_OK;
+            }
+            else if (prompt_return == MINI_INPUT_RET_BACK)
+            {
+                state_machine--;
```

Comment:
there's a case we need to take into account: password is empty. In that case you'll need to go -2